src/potdata/samplers.py

```diff
@@ -616,3 +616,245 @@ def plot2(self, separation1: int, separation2: int, show: bool = False,
+    ) -> list[np.ndarray]:
+        """Select SOAP vectors for atoms of specific species."""
+
+        def select_one(struct, vec):
+            indices = [i for i, s in enumerate(struct.species) if s.symbol in species]
+            return vec[indices]
+
+        selected = []
+        for struct, vec in zip(structures, vectors):
+            selected.append(select_one(struct, vec))
+
+        return selected
+
+    @staticmethod
+    def _get_soap_vector_structure(vectors: list[np.ndarray], pool: str) -> np.ndarray:
+        """Convert SOAP vectors of atoms to SOAP vectors of structures.
+
+        Return a 2D array of shape (n_structures, n_features).
+        """
+        if pool.lower() == "mean":
+            return np.asarray([np.mean(x, axis=0) for x in vectors])
+        elif pool.lower() == "concatenate":
+            return np.asarray([np.ravel(x) for x in vectors])
+        else:
+            supported = ["mean", "concatenate"]
+            raise ValueError(
+                f"Unsupported pooling method `{pool}`. Expected one of " f"{supported}."
+            )
+
+    @staticmethod
+    def _dim_reduction(x: np.ndarray, dim: int | float):
+        """Perform dimension reduction on a 2D array."""
+        pca = PCA(n_components=dim)
+        reduced = pca.fit_transform(x)
+
+        return reduced
+
+    def _cluster_kmeans(self, data: np.ndarray) -> list[int]:
+        """
+        Perform KMeans clustering and return cluster labels.
+        """
+
+        clustering = KMeans(n_clusters=self.kmeans_kwargs["n_clusters"], n_init=10).fit(data)
+        labels = clustering.labels_
+
+        unique_clusters = set(labels)
+        if len(unique_clusters) < self.kmeans_kwargs["n_clusters"]:
+            warnings.warn(
+                f"Number of clusters found ({len(unique_clusters)}) is less than "
+                f"specified n_clusters ({self.kmeans_kwargs['n_clusters']}). "
+                "Consider adjusting the parameters of KMeans."
+            )
+
+        return labels
+
+    def _select_kmeans(
+        self, data: list[Structure], cluster_labels: list[int], ratio: float
+    ) -> tuple[list[int], list[Structure]]:
+        """Select a subset of structures and indices for KMeans."""
+        unique_clusters = set(cluster_labels)
+        sample_size_per_cluster = int(ratio * len(unique_clusters))
+
+        selected_indices = []
+        selected_clusters = []
+
+        for cluster in unique_clusters:
+            cluster_indices = [i for i, label in enumerate(cluster_labels) if label == cluster]
+            cluster_size = min(sample_size_per_cluster, len(cluster_indices))
+            selected_indices.extend(np.random.choice(cluster_indices, cluster_size, replace=False))
+
+        selected_structures = [data[i] for i in selected_indices]
+
+        return selected_indices, selected_structures
+
+    def plot3(self, show: bool = False, figname: str = "kmeans_sample.pdf"):
```

Comment:
Why `plot3`, instead of `plot`? 

Rename it.

Reply:
Because in class DBSCANSamplerStructure, there are already "def plot" and "def plot2", so I thought it would be better if I use "def plot3" to differentiate them

Reply:
need to rename to `plot`